File: android/app/src/main/java/il/co/simplevision/timetrack/data/GreenInvoiceClient.kt
package il.co.simplevision.timetrack.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.time.LocalDate
import java.time.format.DateTimeFormatter

class GreenInvoiceClient(
    private val apiKey: String,
    private val apiSecret: String,
    private val baseUrl: String = "https://api.greeninvoice.co.il/api/v1",
    private val http: OkHttpClient = OkHttpClient(),
    private val json: Json = Json { ignoreUnknownKeys = true; encodeDefaults = true },
) {
    suspend fun createProformaInvoice(
        subject: String,
        itemDescription: String,
        quantity: Double,
        unitPrice: Double,
        vatType: VatType,
        clientId: String,
        date: LocalDate,
    ): String? {
        val token = fetchToken()
        val payload = GreenInvoiceDocumentRequest(
            description = subject,
            client = GreenInvoiceClientReference(id = clientId, accountingKey = null),
            type = 300,
            date = date.format(DATE_FORMAT),
            lang = "en",
            currency = "USD",
            vatType = vatType.rawValue,
            income = listOf(
                GreenInvoiceIncomeItem(
                    description = itemDescription,
                    quantity = quantity,
                    price = unitPrice,
                    currency = "USD",
                    currencyRate = 1.0,
                    vatType = vatType.rawValue,
                ),
            ),
        )

        val url = "$baseUrl/documents"
        val req = Request.Builder()
            .url(url)
            .post(json.encodeToString(payload).toRequestBody(JSON))
            .header("Authorization", "Bearer $token")
            .header("Content-Type", "application/json")
            .build()

        return withContext(Dispatchers.IO) {
            http.newCall(req).execute().use { res ->
                val body = res.body?.string().orEmpty()
                if (!res.isSuccessful) throw GreenInvoiceError.Server(res.code, body.ifBlank { null })
                runCatching { json.decodeFromString(GreenInvoiceResponse.serializer(), body) }.getOrNull()
                    ?.let { it.id ?: it.documentId ?: it.number }
            }
        }
    }

    suspend fun fetchClientId(accountingKey: String): String? {
        val trimmed = accountingKey.trim()
        if (trimmed.isEmpty()) return null
        val token = fetchToken()
        return fetchClientIdInternal(trimmed, token)
    }

    private suspend fun fetchToken(): String {
        val payload = GreenInvoiceAuthRequest(id = apiKey, secret = apiSecret)
        val url = "$baseUrl/account/token"
        val req = Request.Builder()
            .url(url)
            .post(json.encodeToString(payload).toRequestBody(JSON))
            .header("Content-Type", "application/json")
            .build()

        return withContext(Dispatchers.IO) {
            http.newCall(req).execute().use { res ->
                val body = res.body?.string().orEmpty()
                if (!res.isSuccessful) throw GreenInvoiceError.Server(res.code, body.ifBlank { null })

                val decoded = runCatching { json.decodeFromString(GreenInvoiceAuthResponse.serializer(), body) }.getOrNull()
                decoded?.tokenValue()?.let { return@use it }

                extractTokenFromJson(body)?.let { return@use it }

                throw GreenInvoiceError.MissingToken
            }
        }
    }

    private suspend fun fetchClientIdInternal(accountingKey: String, token: String): String? {
        val queryKeys = listOf("accountingKey", "search", "query", "q")
        val postPayloads: List<Map<String, Any?>> = listOf(
            mapOf("search" to accountingKey),
            mapOf("query" to accountingKey),
            mapOf("q" to accountingKey),
            mapOf("accountingKey" to accountingKey),
            mapOf("accountingNumber" to accountingKey),
            mapOf("customerKey" to accountingKey),
            mapOf("externalId" to accountingKey),
            mapOf("filter" to mapOf("accountingKey" to accountingKey)),
            mapOf("client" to mapOf("accountingKey" to accountingKey)),
        )

        var lastError: Throwable? = null
        var hadResponse = false

        for (payload in postPayloads) {
            try {
                val result = fetchClientId(
                    path = "clients/search",
                    token = token,
                    method = "POST",
                    query = null,
                    payload = payload,
                    accountingKey = accountingKey,
                )
                hadResponse = true
                if (result != null) return result
            } catch (t: Throwable) {
                if (t is GreenInvoiceError.Server && t.statusCode == 405) continue
                lastError = t
            }
        }

        for (queryKey in queryKeys) {
            try {
                val result = fetchClientId(
                    path = "clients/search",
                    token = token,
                    method = "GET",
                    query = mapOf(queryKey to accountingKey),
                    payload = null,
                    accountingKey = accountingKey,
                )
                hadResponse = true
                if (result != null) return result
            } catch (t: Throwable) {
                if (t is GreenInvoiceError.Server && t.statusCode == 405) continue
                lastError = t
            }
        }

        for (queryKey in queryKeys) {
            try {
                val result = fetchClientId(
                    path = "clients",
                    token = token,
                    method = "GET",
                    query = mapOf(queryKey to accountingKey),
                    payload = null,
                    accountingKey = accountingKey,
                )
                hadResponse = true
                if (result != null) return result
            } catch (t: Throwable) {
                if (t is GreenInvoiceError.Server && t.statusCode == 405) continue
                lastError = t
            }
        }

        if (hadResponse) return null
        if (lastError != null) throw lastError
        return null
    }

    private suspend fun fetchClientId(
        path: String,
        token: String,
        method: String,
        query: Map<String, String>?,
        payload: Map<String, Any?>?,
        accountingKey: String,
    ): String? {
        val url = buildUrl("$baseUrl/$path", query)
        val builder = Request.Builder()
            .url(url)
            .header("Authorization", "Bearer $token")

        if (method == "POST") {
            val bodyJson = mapToJsonString(payload ?: emptyMap())
            val body = bodyJson.toRequestBody(JSON)
            builder.post(body).header("Content-Type", "application/json")
        } else {
            builder.get()
        }

        val req = builder.build()
        return withContext(Dispatchers.IO) {
            http.newCall(req).execute().use { res ->
                val body = res.body?.string().orEmpty()
                if (!res.isSuccessful) throw GreenInvoiceError.Server(res.code, body.ifBlank { null })
                extractClientId(body, accountingKey)
            }
        }
    }

    private fun buildUrl(base: String, query: Map<String, String>?): String {
        if (query.isNullOrEmpty()) return base
        val encoded = query.entries.joinToString("&") { (k, v) ->
            val ek = URLEncoder.encode(k, StandardCharsets.UTF_8.name())
            val ev = URLEncoder.encode(v, StandardCharsets.UTF_8.name())
            "$ek=$ev"
        }
        return "$base?$encoded"
    }

    private fun mapToJsonString(map: Map<String, Any?>): String {
        val el = anyToJsonElement(map)
        return json.encodeToString(JsonElement.serializer(), el)
    }

    private fun anyToJsonElement(value: Any?): JsonElement {
        return when (value) {
            null -> JsonNull
            is JsonElement -> value
            is String -> JsonPrimitive(value)
            is Number -> JsonPrimitive(value)
            is Boolean -> JsonPrimitive(value)
            is Map<*, *> -> {
                val content = LinkedHashMap<String, JsonElement>()
                value.entries.forEach { (k, v) ->
                    val key = k?.toString() ?: return@forEach
                    content[key] = anyToJsonElement(v)
                }
                JsonObject(content)
            }
            is List<*> -> JsonArray(value.map { anyToJsonElement(it) })
            else -> JsonPrimitive(value.toString())
        }
    }

    private fun extractClientId(body: String, accountingKey: String): String? {
        val normalizedKey = accountingKey.trim().lowercase()
        val root = runCatching { json.parseToJsonElement(body) }.getOrNull() ?: return null

        fun stringValue(obj: JsonObject, keys: List<String>): String? {
            for (key in keys) {
                val el = obj[key] ?: continue
                if (el is JsonPrimitive) {
                    if (el.isString) {
                        val s = el.content
                        if (s.isNotBlank()) return s
                    } else {
                        val n = el.content
                        if (n.isNotBlank()) return n
                    }
                }
            }
            return null
        }

        fun findClientId(clients: List<JsonObject>): String? {
            val matchKeys = listOf(
                "accountingKey", "accountingCode", "accountingId", "accountingNumber",
                "customerKey", "clientKey", "externalId", "externalID", "key", "code",
            )
            for (client in clients) {
                for (key in matchKeys) {
                    val value = stringValue(client, listOf(key))?.lowercase() ?: continue
                    if (value == normalizedKey) {
                        return stringValue(client, listOf("id", "clientId", "clientID", "_id"))
                    }
                }
            }
            if (clients.size == 1) {
                return stringValue(clients[0], listOf("id", "clientId", "clientID", "_id"))
            }
            return null
        }

        if (root is JsonObject) {
            stringValue(root, listOf("id", "clientId", "clientID", "_id"))?.let { return it }
            (root["client"] as? JsonObject)?.let { client ->
                stringValue(client, listOf("id", "clientId", "clientID", "_id"))?.let { return it }
            }
            for (key in listOf("items", "data", "clients", "results")) {
                val el = root[key] ?: continue
                if (el is JsonArray) {
                    val clients = el.mapNotNull { it as? JsonObject }
                    findClientId(clients)?.let { return it }
                }
            }
            return null
        }

        if (root is JsonArray) {
            val clients = root.mapNotNull { it as? JsonObject }
            return findClientId(clients)
        }

        return null
    }

    private fun extractTokenFromJson(body: String): String? {
        val root = runCatching { json.parseToJsonElement(body) }.getOrNull() ?: return null
        fun findToken(el: JsonElement): String? {
            if (el is JsonObject) {
                for (key in listOf("token", "access_token", "accessToken")) {
                    val v = el[key]
                    if (v is JsonPrimitive && v.content.isNotBlank()) return v.content
                }
                val nested = el["data"]
                if (nested != null) return findToken(nested)
            }
            return null
        }
        return findToken(root)
    }

    companion object {
        val sandboxBaseUrl: String = "https://sandbox.d.greeninvoice.co.il/api/v1"
        private val JSON = "application/json; charset=utf-8".toMediaType()
        private val DATE_FORMAT: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
    }
}

@Serializable
private data class GreenInvoiceDocumentRequest(
    val description: String,
    val client: GreenInvoiceClientReference? = null,
    val type: Int,
    val date: String,
    val lang: String,
    val currency: String,
    val vatType: Int,
    val income: List<GreenInvoiceIncomeItem>,
)

@Serializable
private data class GreenInvoiceClientReference(
    val id: String? = null,
    val accountingKey: String? = null,
)

@Serializable
private data class GreenInvoiceIncomeItem(
    val description: String,
    val quantity: Double,
    val price: Double,
    val currency: String,
    val currencyRate: Double,
    val vatType: Int,
)

@Serializable
private data class GreenInvoiceAuthRequest(
    val id: String,
    val secret: String,
)

@Serializable
private data class GreenInvoiceAuthResponse(
    val token: String? = null,
    @SerialName("accessToken") val accessToken: String? = null,
    @SerialName("access_token") val accessTokenSnake: String? = null,
    val data: TokenContainer? = null,
) {
    @Serializable
    data class TokenContainer(
        val token: String? = null,
        @SerialName("accessToken") val accessToken: String? = null,
        @SerialName("access_token") val accessTokenSnake: String? = null,
    )

    fun tokenValue(): String? {
        return token
            ?: accessToken
            ?: accessTokenSnake
            ?: data?.token
            ?: data?.accessToken
            ?: data?.accessTokenSnake
    }
}

@Serializable
private data class GreenInvoiceResponse(
    val id: String? = null,
    val documentId: String? = null,
    val number: String? = null,
)

sealed class GreenInvoiceError(message: String) : Exception(message) {
    object InvalidResponse : GreenInvoiceError("Invalid response from GreenInvoice.")
    data class Server(val statusCode: Int, val body: String?) : GreenInvoiceError(
        if (!body.isNullOrBlank()) "GreenInvoice error $statusCode: $body" else "GreenInvoice error $statusCode.",
    )

    object MissingToken : GreenInvoiceError("Could not read an authentication token from GreenInvoice.")
}

File: Timetrack/GreenInvoiceClient.swift
import Foundation

struct GreenInvoiceClient {
    let apiKey: String
    let apiSecret: String
    let baseURL: URL
    private let session: URLSession = .shared

    static let sandboxBaseURL = URL(string: "https://sandbox.d.greeninvoice.co.il/api/v1")!

    init(
        apiKey: String,
        apiSecret: String,
        baseURL: URL = URL(string: "https://api.greeninvoice.co.il/api/v1")!
    ) {
        self.apiKey = apiKey
        self.apiSecret = apiSecret
        self.baseURL = baseURL
    }

    func createProformaInvoice(
        subject: String,
        itemDescription: String,
        quantity: Double,
        unitPrice: Double,
        vatType: VatType,
        clientId: String,
        date: Date
    ) async throws -> String? {
        let token = try await fetchToken()
        let documentType = 300
        let currency = "USD"
        let payload = GreenInvoiceDocumentRequest(
            description: subject,
            client: GreenInvoiceClientReference(id: clientId, accountingKey: nil),
            type: documentType,
            date: Self.dateFormatter.string(from: date),
            lang: "en",
            currency: currency,
            vatType: vatType.rawValue,
            income: [
                GreenInvoiceIncomeItem(
                    description: itemDescription,
                    quantity: quantity,
                    price: unitPrice,
                    currency: currency,
                    currencyRate: 1,
                    vatType: vatType.rawValue
                )
            ]
        )

        var request = URLRequest(url: baseURL.appendingPathComponent("documents"))
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(payload)

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw GreenInvoiceError.invalidResponse
        }
        guard (200..<300).contains(httpResponse.statusCode) else {
            let message = String(data: data, encoding: .utf8)
            throw GreenInvoiceError.server(statusCode: httpResponse.statusCode, message: message)
        }
        if let decoded = try? JSONDecoder().decode(GreenInvoiceResponse.self, from: data) {
            return decoded.id ?? decoded.documentId ?? decoded.number
        }
        return nil
    }

    func fetchClientId(accountingKey: String) async throws -> String? {
        let trimmed = accountingKey.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        let token = try await fetchToken()
        return try await fetchClientId(accountingKey: trimmed, token: token)
    }

    private func fetchToken() async throws -> String {
        let payload = GreenInvoiceAuthRequest(id: apiKey, secret: apiSecret)
        var request = URLRequest(url: baseURL.appendingPathComponent("account/token"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(payload)

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw GreenInvoiceError.invalidResponse
        }
        guard (200..<300).contains(httpResponse.statusCode) else {
            let message = String(data: data, encoding: .utf8)
            throw GreenInvoiceError.server(statusCode: httpResponse.statusCode, message: message)
        }

        if let decoded = try? JSONDecoder().decode(GreenInvoiceAuthResponse.self, from: data),
           let token = decoded.tokenValue {
            return token
        }

        if let token = extractToken(from: data) {
            return token
        }

        throw GreenInvoiceError.missingToken
    }

    private func fetchClientId(accountingKey: String, token: String) async throws -> String? {
        let queryKeys = ["accountingKey", "search", "query", "q"]
        let postPayloads: [[String: Any]] = [
            ["search": accountingKey],
            ["query": accountingKey],
            ["q": accountingKey],
            ["accountingKey": accountingKey],
            ["accountingNumber": accountingKey],
            ["customerKey": accountingKey],
            ["externalId": accountingKey],
            ["filter": ["accountingKey": accountingKey]],
            ["client": ["accountingKey": accountingKey]]
        ]

        var lastError: Error?
        var hadResponse = false

        for payload in postPayloads {
            do {
                let result = try await fetchClientId(
                    path: "clients/search",
                    token: token,
                    method: "POST",
                    queryItems: nil,
                    payload: payload,
                    accountingKey: accountingKey
                )
                hadResponse = true
                if let id = result {
                    return id
                }
            } catch {
                if isMethodNotAllowed(error) {
                    continue
                }
                lastError = error
            }
        }

        for queryKey in queryKeys {
            do {
                let result = try await fetchClientId(
                    path: "clients/search",
                    token: token,
                    method: "GET",
                    queryItems: [URLQueryItem(name: queryKey, value: accountingKey)],
                    payload: nil,
                    accountingKey: accountingKey
                )
                hadResponse = true
                if let id = result {
                    return id
                }
            } catch {
                if isMethodNotAllowed(error) {
                    continue
                }
                lastError = error
            }
        }

        for queryKey in queryKeys {
            do {
                let result = try await fetchClientId(
                    path: "clients",
                    token: token,
                    method: "GET",
                    queryItems: [URLQueryItem(name: queryKey, value: accountingKey)],
                    payload: nil,
                    accountingKey: accountingKey
                )
                hadResponse = true
                if let id = result {
                    return id
                }
            } catch {
                if isMethodNotAllowed(error) {
                    continue
                }
                lastError = error
            }
        }

        if hadResponse {
            return nil
        }
        if let lastError {
            throw lastError
        }
        return nil
    }

    private func fetchClientId(
        path: String,
        token: String,
        method: String,
        queryItems: [URLQueryItem]?,
        payload: [String: Any]?,
        accountingKey: String
    ) async throws -> String? {
        var url = baseURL.appendingPathComponent(path)
        if let queryItems, !queryItems.isEmpty {
            var components = URLComponents(url: url, resolvingAgainstBaseURL: false)
            components?.queryItems = queryItems
            if let composed = components?.url {
                url = composed
            }
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        if let payload {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: payload, options: [])
        }

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw GreenInvoiceError.invalidResponse
        }
        guard (200..<300).contains(httpResponse.statusCode) else {
            let message = String(data: data, encoding: .utf8)
            throw GreenInvoiceError.server(statusCode: httpResponse.statusCode, message: message)
        }

        return extractClientId(from: data, accountingKey: accountingKey)
    }

    private func isMethodNotAllowed(_ error: Error) -> Bool {
        if let greenError = error as? GreenInvoiceError {
            if case let .server(statusCode, _) = greenError, statusCode == 405 {
                return true
            }
        }
        return false
    }

    private func extractClientId(from data: Data, accountingKey: String) -> String? {
        guard let json = try? JSONSerialization.jsonObject(with: data) else { return nil }
        let normalizedKey = accountingKey.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()

        if let dict = json as? [String: Any] {
            if let id = stringValue(from: dict, keys: ["id", "clientId", "clientID", "_id"]) {
                return id
            }
            if let client = dict["client"] as? [String: Any],
               let id = stringValue(from: client, keys: ["id", "clientId", "clientID", "_id"]) {
                return id
            }
            for key in ["items", "data", "clients", "results"] {
                if let array = dict[key] as? [[String: Any]],
                   let id = findClientId(in: array, accountingKey: normalizedKey) {
                    return id
                }
            }
            return nil
        }

        if let array = json as? [[String: Any]] {
            return findClientId(in: array, accountingKey: normalizedKey)
        }

        return nil
    }

    private func findClientId(in clients: [[String: Any]], accountingKey: String) -> String? {
        let matchKeys = [
            "accountingKey", "accountingCode", "accountingId", "accountingNumber",
            "customerKey", "clientKey", "externalId", "externalID", "key", "code"
        ]

        for client in clients {
            for key in matchKeys {
                if let value = stringValue(from: client, keys: [key])?.lowercased(),
                   value == accountingKey {
                    return stringValue(from: client, keys: ["id", "clientId", "clientID", "_id"])
                }
            }
        }

        if clients.count == 1 {
            return stringValue(from: clients[0], keys: ["id", "clientId", "clientID", "_id"])
        }

        return nil
    }

    private func stringValue(from dict: [String: Any], keys: [String]) -> String? {
        for key in keys {
            if let value = dict[key] {
                if let string = value as? String, !string.isEmpty {
                    return string
                }
                if let number = value as? NSNumber {
                    return number.stringValue
                }
            }
        }
        return nil
    }

    private func extractToken(from data: Data) -> String? {
        guard let json = try? JSONSerialization.jsonObject(with: data) else { return nil }
        if let dict = json as? [String: Any] {
            return findToken(in: dict)
        }
        return nil
    }

    private func findToken(in dict: [String: Any]) -> String? {
        for key in ["token", "access_token", "accessToken"] {
            if let value = dict[key] as? String, !value.isEmpty {
                return value
            }
        }
        if let nested = dict["data"] as? [String: Any] {
            return findToken(in: nested)
        }
        return nil
    }

    private static let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone.current
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()
}

struct GreenInvoiceDocumentRequest: Encodable {
    let description: String
    let client: GreenInvoiceClientReference?
    let type: Int
    let date: String
    let lang: String
    let currency: String
    let vatType: Int
    let income: [GreenInvoiceIncomeItem]
}

struct GreenInvoiceClientReference: Encodable {
    let id: String?
    let accountingKey: String?
}

struct GreenInvoiceIncomeItem: Encodable {
    let description: String
    let quantity: Double
    let price: Double
    let currency: String
    let currencyRate: Double
    let vatType: Int
}

struct GreenInvoiceAuthRequest: Encodable {
    let id: String
    let secret: String
}

struct GreenInvoiceAuthResponse: Decodable {
    let token: String?
    let accessToken: String?
    let data: TokenContainer?

    var tokenValue: String? {
        token ?? accessToken ?? data?.token ?? data?.accessToken
    }

    struct TokenContainer: Decodable {
        let token: String?
        let accessToken: String?
    }
}

struct GreenInvoiceResponse: Decodable {
    let id: String?
    let documentId: String?
    let number: String?
}

enum GreenInvoiceError: LocalizedError {
    case invalidResponse
    case server(statusCode: Int, message: String?)
    case missingToken

    var errorDescription: String? {
        switch self {
        case .invalidResponse:
            return "Invalid response from GreenInvoice."
        case let .server(statusCode, message):
            if let message, !message.isEmpty {
                return "GreenInvoice error \(statusCode): \(message)"
            }
            return "GreenInvoice error \(statusCode)."
        case .missingToken:
            return "Could not read an authentication token from GreenInvoice."
        }
    }
}
